File: lib/tps/repo/conversation.ex
defmodule TPS.Repo.Conversation do
  def new_convo(conn, users) do
    {:ok, statement} =
      conn
      |> Exqlite.Sqlite3.prepare("insert into conversation values (?1, ?2)")

    id = UUID.uuid4()

    :ok = Exqlite.Sqlite3.bind(statement, [id, users])

    case Exqlite.Sqlite3.step(conn, statement) do
      {:error, reason} ->
        Exqlite.Sqlite3.release(conn, statement)
        {:error, reason}

      :done ->
        Exqlite.Sqlite3.release(conn, statement)
        {:ok, id}
    end
  end

  def conversations_by_username(conn, username) do
    {:ok, statement} =
      conn
      |> Exqlite.Sqlite3.prepare("select * from conversation where users like '%?1%'")

    :ok = Exqlite.Sqlite3.bind(statement, [username])

    case Exqlite.Sqlite3.multi_step(conn, statement) do
      {:error, reason} ->
        Exqlite.Sqlite3.release(conn, statement)
        {:error, reason}

      {:rows, result} ->
        Exqlite.Sqlite3.release(conn, statement)
        {:ok, result}
    end
  end

  def conversation_users(conn, id) do
    {:ok, statement} =
      conn
      |> Exqlite.Sqlite3.prepare("select users from conversation where id=?1")

    :ok = Exqlite.Sqlite3.bind(statement, [id])

    case Exqlite.Sqlite3.step(conn, statement) do
      {:error, reason} ->
        Exqlite.Sqlite3.release(conn, statement)
        {:error, reason}

      {:row, result} ->
        Exqlite.Sqlite3.release(conn, statement)
        {:ok, result}
    end
  end
end

File: internal/models/conversation.go
package models

import (
	"log"

	"github.com/Robert-Duck-by-BB-SR/tps/internal/database"
)

type Conversation struct {
	Id, Users string
}

func FetchConversationUsers(id string) (error, string) {
	var users string
	if err := database.DB.Get(&users, "select users from conversation where id=?", id); err != nil {
		log.Println("could not fetch users from conversation: ", err)
		return err, ""
	}
	return nil, users
}

func FetchConversationsByUsername(username string) (error, []Conversation) {
	var conversations []Conversation
	if err := database.DB.Select(&conversations, "select * from conversation where users like '%?%'", username); err != nil {
		log.Println("could not fetch users from conversation: ", err)
		return err, []Conversation{}
	}
	return nil, conversations
}

func CreateConversation(id, users string) error {
	if _, err := database.DB.Exec("insert into conversation values (?, ?)", id, users); err != nil {
		return err
	}
	return nil
}
